File: tests/bench/strings.go
// The same work as `strings.skuld`, in Go. See `README.md`.
package main

import (
	"fmt"
	"os"
	"strconv"
	"unicode/utf8"
)

func pushNumber(buffer []byte, value int) []byte {
	digits := []byte{}
	rest := value
	if rest == 0 {
		digits = append(digits, 48)
	}
	for rest > 0 {
		digits = append(digits, byte(48+rest%10))
		rest /= 10
	}
	for index := len(digits) - 1; index >= 0; index-- {
		buffer = append(buffer, digits[index])
	}
	return buffer
}

func main() {
	count := 0
	if len(os.Args) > 1 {
		count, _ = strconv.Atoi(os.Args[1])
	}
	buffer := []byte{}
	for index := 0; index < count; index++ {
		buffer = append(buffer, []byte("item-")...)
		buffer = pushNumber(buffer, index)
		buffer = append(buffer, 59)
	}
	// Validated, like Skuld's `bytes_to_string`.
	if !utf8.Valid(buffer) {
		fmt.Println("not utf-8")
		os.Exit(1)
	}
	text := string(buffer)
	found := 0
	for at := 0; at+1 < len(text); at++ {
		if text[at] == 57 && text[at+1] == 59 {
			found++
		}
	}
	fmt.Println(len(text), found)
}

File: tests/bench/strings.rs
// The same work as `strings.skuld`, in Rust. See `README.md`.

fn push_number(buffer: &mut Vec<u8>, value: i64) {
    let mut digits: Vec<u8> = Vec::new();
    let mut rest = value;
    if rest == 0 {
        digits.push(48);
    }
    while rest > 0 {
        digits.push(48 + (rest % 10) as u8);
        rest /= 10;
    }
    for index in (0..digits.len()).rev() {
        buffer.push(digits[index]);
    }
}

fn main() {
    let count: i64 = std::env::args()
        .nth(1)
        .and_then(|text| text.parse().ok())
        .unwrap_or(0);
    let mut buffer: Vec<u8> = Vec::new();
    for index in 0..count {
        buffer.extend_from_slice(b"item-");
        push_number(&mut buffer, index);
        buffer.push(59);
    }
    // Validated, like Skuld's `bytes_to_string`: the check is part of the work.
    let text = String::from_utf8(buffer).expect("utf-8");
    let bytes = text.as_bytes();
    let mut found = 0;
    let mut at = 0;
    while at + 1 < bytes.len() {
        if bytes[at] == 57 && bytes[at + 1] == 59 {
            found += 1;
        }
        at += 1;
    }
    println!("{} {}", text.len(), found);
}
